lib/filter-schema.js

```diff
@@ -0,0 +1,163 @@
+      context)
+
+    return {
+      schema: filteredSchema
+    }
+  }
+}
+
+function isIntrospection (document) {
+  for (const queryType of document.definitions) {
+    if (queryType.operation !== 'query') {
+      // if there is a mutation or subscription, we can skip the introspection check
+      break
+    }
+
+    // if there is an introspection operation, we must filter the schema
+    if (queryType.selectionSet.selections.some(sel => (
+      sel.name.value === '__schema' ||
+      sel.name.value === '__type'
+    ))) {
+      return true
+    }
+  }
+  return false
+}
+
+async function filterSchema (graphQLSchema, policies, context) {
+  const filterDirectiveMap = {}
+
+  // each `policies` item is a directive
+  let skipFiltering = true
+  for (const { policy, policyFunction } of policies) {
+    // each `policy` contains all the GraphQL OBJECT and FIELDS that are affected by the directive
+    for (const [typeName, typePolicy] of Object.entries(policy)) {
+      // different `policies` item can affect the same GraphQL OBJECT
+      if (filterDirectiveMap[typeName] === undefined) {
+        filterDirectiveMap[typeName] = {}
+      } else if (filterDirectiveMap[typeName] === false) {
+        // if the object has already been filtered, we can skip the field processing
+        continue
+      }
+
+      const schemaType = graphQLSchema.getType(typeName)
+      const schemaTypeFields = typeof schemaType.getFields === 'function'
+        ? schemaType.getFields()
+        : {}
+      for (const [fieldName, fieldPolicy] of Object.entries(typePolicy)) {
+        // each `fieldName` is a single GraphQL item associated with the directive
+
+        if (filterDirectiveMap[typeName] === false || filterDirectiveMap[typeName][fieldName] === false) {
+          // if we have already decided to filter out this field
+          // it does not need to be checked again
+          continue
+        }
+
+        let canShowDirectiveField = true
+        const isObjectPolicy = fieldName === '__typePolicy'
+        try {
+          // https://github.com/graphql/graphql-js/blob/main/src/type/definition.ts#L974
+          const info = {
+            fieldName,
+            fieldNodes: schemaType.astNode.fields,
+            returnType: isObjectPolicy ? '' : schemaTypeFields[fieldName].type, // TODO
+            parentType: schemaType,
+            schema: graphQLSchema,
+            fragments: {},
+            rootValue: {},
+            operation: { kind: 'OperationDefinition', operation: 'query' },
+            variableValues: {}
+          }
+          // The undefined parameters are: https://graphql.org/learn/execution/#root-fields-resolvers
+          // - parent: it is not possible know it since the resolver is not executed yet
+          // - args: it is not expected that the introspection query will have arguments for the directives policies
+          canShowDirectiveField = await policyFunction(fieldPolicy, undefined, undefined, context, info)
```

Comment:
what happens if this throws an error? Also, could we set `parent`, `args` explicitly to `null`?

Reply:
> what happens if this throws an error?

It is considered as a failed check - we can add an option for this if needed

> Also, could we set parent, args explicitly to null?

Done